crates/solvers/src/util/num.rs

```diff
@@ -0,0 +1,63 @@
+//! Numeric and arithmetic utilities.
+
+use {
+    bigdecimal::BigDecimal,
+    num::{BigInt, Signed},
+};
+
+/// Round a [`bigdecimal::BigDecimal`] to the specified precision.
+///
+/// This is the same as [`bigdecimal::BigDecimal::round`], but it does not panic
+/// for high precision values.
+pub fn round(x: &BigDecimal, round_digits: i64) -> BigDecimal {
```

Comment:
Instead of reimplementing the original `round` couldn't you just detect the case where it would panic and call `with_prec(round_digits + 1)` to avoid the panic. I don't know if this works but I think it's probably less error prone to build the same functionality by composing existing functions.

Reply:
I think this is a good suggestion!

The reason I originally dismissed this was that `with_prec` and `round`-ing have fundamentally different semantics:

```rust
let b = "1234.56789".parse::<BigDecimal>().unwrap();
println!("with_prec: {}", b.with_prec(2));
println!("round:     {}", b.round(2));
```
```
with_prec: 1200
round:     1234.57
```

that being said, we **should** have the invariant that slippage is less than 1 since the limits has the invariant that `slippage.relative ∈ [0, 1]`. This means that we should be able to do it the way you suggested.

A slightly more "fool-proof" approach would be to use `.with_prec(MAX_PRECISION_BEFORE_ROUND_PANICS)`. Computing `MAX_PRECISION_BEFORE_ROUND_PANICS` is actually kind of straight forward - in the `round` implementation, it basically does a `BigInt.try_into::<i128>().unwrap()` which causes the panic for very large precision `BigDecimal` values. This means that `i128::MAX.to_string().len() - 1` is the maximum allowed precision for use with `round`. In fact, you can test this theory and see that it seems to be correct:

```rust
let max = i128::MAX.to_string().len() - 1;
println!("MAX = {max}");
for i in 1.. {
    println!("{i:>2}: {}", b.with_prec(i).round(4));
}
```

Yields:
```
MAX = 38
 1: 40
...
38: 42.1158
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', /Users/nlordell/.cargo/registry/src/github.com-1ecc6299db9ec823/bigdecimal-0.3.0/src/lib.rs:594:43
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

So, a pretty pragmatic work around for this would be `slippage.with_prec(38).round(i)`. WDYT?

Reply:
LGTM

Reply:
26ceb91aa1830092822e6b2573075b6a973aa6d5